statsmodels/tsa/stattools.py

```diff
@@ -337,32 +351,64 @@ def acovf(x, unbiased=False, demean=True, fft=False, missing='none'):
     if deal_with_masked:
         if missing == 'raise':
             raise MissingDataError("NaNs were encountered in the data")
-        notmask_bool = ~np.isnan(x) #bool
+        notmask_bool = ~np.isnan(x)  # bool
         if missing == 'conservative':
+            # Must copy for thread safety
+            x = x.copy()
             x[~notmask_bool] = 0
-        else: #'drop'
-            x = x[notmask_bool] #copies non-missing
-        notmask_int = notmask_bool.astype(int) #int
+        else:  # 'drop'
+            x = x[notmask_bool]  # copies non-missing
+        notmask_int = notmask_bool.astype(int)  # int
 
     if demean and deal_with_masked:
         # whether 'drop' or 'conservative':
-        xo = x - x.sum()/notmask_int.sum()
-        if missing=='conservative':
+        xo = x - x.sum() / notmask_int.sum()
+        if missing == 'conservative':
             xo[~notmask_bool] = 0
     elif demean:
         xo = x - x.mean()
     else:
         xo = x
 
     n = len(x)
-    if unbiased and deal_with_masked and missing=='conservative':
+    lag_len = nlag
+    if nlag is None:
+        lag_len = n - 1
+    elif nlag > n - 1:
+        raise ValueError('nlag must be smaller than nobs - 1')
+
+    if not fft and nlag is not None:
+        acov = np.empty(lag_len + 1)
+        acov[0] = xo.dot(xo)
+        for i in range(lag_len):
+            acov[i + 1] = xo[i + 1:].dot(xo[:-(i + 1)])
+        if not deal_with_masked or missing == 'drop':
+            if unbiased:
+                acov /= (n - np.arange(lag_len + 1))
+            else:
+                acov /= n
+        else:
+            if unbiased:
+                divisor = np.empty(lag_len + 1, dtype=np.int64)
+                divisor[0] = notmask_int.sum()
+                for i in range(lag_len):
+                    divisor[i + 1] = notmask_int[i + 1:].dot(notmask_int[:-(i + 1)])
+                    # divisor[i + 1] = [notmask_int[i + 1:] * notmask_int[:-(i + 1)].sum()
+                divisor[divisor == 0] = 1
+                acov /= divisor
+            else:
+                acov /= notmask_int.sum()
+        return acov
+
```

Comment:
add a comment line here, about the case that's left "either fft or nlag is None", i.e. when we use correlate or fft.
When reading through the function, the return in this path is not obvious enough.